components/places/src/import/ios_bookmarks.rs

```diff
@@ -0,0 +1,566 @@
+/// in question.
+///
+/// ### Basic process
+///
+/// - Attach the iOS database.
+/// - Slurp records into a temp table "iosBookmarksStaging" from iOS database.
+/// - Add any entries to moz_places that are needed (in practice, they'll all be
+///   needed, we don't yet store history for iOS).
+/// - Fill mirror using iosBookmarksStaging.
+/// - Fill mirror with tags from iosBookmarksStaging.
+/// - Fill mirror structure using both iOS database and iosBookmarksStaging.
+/// - Run dogear merge
+/// - Use iosBookmarksStaging to fixup the data that was actually inserted.
+/// - Update frecency for new items.
+/// - Cleanup (Delete mirror and mirror structure, detach iOS database, etc).
+pub fn import_ios_bookmarks(
+    places_api: &PlacesApi,
+    path: impl AsRef<std::path::Path>,
+) -> Result<()> {
+    let url = crate::util::ensure_url_path(path)?;
+    do_import_ios_bookmarks(places_api, url)
+}
+
+fn do_import_ios_bookmarks(places_api: &PlacesApi, ios_db_file_url: Url) -> Result<()> {
+    let conn = places_api.open_sync_connection()?;
+
+    let scope = conn.begin_interrupt_scope();
+
+    sql_fns::define_functions(&conn)?;
+
+    // Not sure why, but apparently beginning a transaction sometimes
+    // fails if we open the DB as read-only. Hopefully we don't
+    // unintentionally write to it anywhere...
+    // ios_db_file_url.query_pairs_mut().append_pair("mode", "ro");
+
+    // Note: `Drop` is executed in FIFO order, so this will happen after
+    // the transaction's drop, which is what we want.
+    let clear_mirror_on_drop = ExecuteOnDrop {
+        conn: &conn,
+        sql: &WIPE_MIRROR,
+    };
+
+    log::trace!("Attaching database {}", ios_db_file_url);
+    let auto_detach = attached_database(&conn, &ios_db_file_url)?;
+
+    let tx = conn.begin_transaction()?;
+
+    // Clear the mirror now, since we're about to fill it with data from the ios
+    // connection.
+    log::debug!("Clearing mirror to prepare for import");
+    conn.execute_batch(&WIPE_MIRROR)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Creating staging table");
+    conn.execute_batch(&CREATE_STAGING_TABLE)?;
+
+    log::debug!("Importing from iOS to staging table");
+    conn.execute_batch(&POPULATE_STAGING)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating missing entries in moz_places");
+    conn.execute_batch(&FILL_MOZ_PLACES)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror");
+    conn.execute_batch(&POPULATE_MIRROR)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror tags");
+    populate_mirror_tags(&conn)?;
+    scope.err_if_interrupted()?;
+
+    // Ideally we could just do this right after `CREATE_AND_POPULATE_STAGING`,
+    // which would mean we could detach the iOS database sooner, but we have
+    // constraints on the mirror structure that prevent this (and there's
+    // probably nothing bad that can happen in this case anyway).
```

Comment:
If you didn't know about `PRAGMA defer_foreign_keys = true`, it'll let you get around the constraints until you're ready to commit the transaction! But I think keeping it attached is completely fine.

Reply:
Oh, that comment is wrong, we can't detach the ios db until after the transaction or things get mad.